Merge pytest-catchlog plugin

thisch opened the pull request:
In this PR the pytest-catchlog plugin is merged (develop branch of the
pytest-catchlog repo) into pytest-core, as
suggested in (eisensheng/pytest-catchlog#8). In the course of that I've renamed the
plugin from CatchLogPlugin to LoggingPlugin (see eisensheng/pytest-catchlog#8).

pytest-catchlog contains a backward compatibilty code for pytest-capturelog,
which I removed in this PR, since I don't think that it makes sense to merge 
this code into pytest-core.  Futhermore, the pytest-catchlog repo contains
performance tests
(https://github.com/eisensheng/pytest-catchlog/tree/develop/tests/perf)
which are not yet part of this PR. Do we want to integrate them? If
yes, how?

The documentation for this PR still needs to be integrated into this
repo. I'll probably copy the docu from the readme of pytest-catchlog and
adapt it a bit.

- [x] Add a new news fragment into the changelog folder
  * name it `$issue_id.$type` for example (588.bug)
  * if you don't have an issue_id change it to the pr id after creating the pr
  * ensure type is one of `removal`, `feature`, `bugfix`, `vendor`, `doc` or `trivial`
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."
- [x] Target: for `bugfix`, `vendor`, `doc` or `trivial` fixes, target `master`; for removals or features target `features`;
- [x] Make sure to include reasonable tests for your change if necessary

Unless your change is a trivial or a documentation fix (e.g.,  a typo or reword of a small section) please:

- [x] Add yourself to `AUTHORS`;

The-Compiler commented:
This should go to the `features` branch, not `master`.

RonnyPfannschmidt commented:
thanks for bringing this forward, i'd love to participate in the finalizing discussions but im off the grid next week

please dont forget to mark provided fixtures/apis as experimental so we can more easily bring in changes that may turn necessary by feedback coming in from wider adoption

thisch commented:
> please dont forget to mark provided fixtures/apis as experimental so we can more easily bring in changes that may turn necessary by feedback coming in from wider adoption

Is it documented somewhere how I can mark the features/apis as experimental? 

See you in a week @RonnyPfannschmidt

RonnyPfannschmidt commented:
@thisch we dont yet have a formal/in code mechanism, for now please document, and lets add a note/issue for FutureWarnings

nicoddemus commented on _pytest/logging.py:
```diff
@@ -0,0 +1,356 @@
+
+    Defaults to the root logger.
+    """
+    if logger is None or isinstance(logger, py.builtin._basestring):
```
You can use `six` instead of `py.builtin`, we are phasing out `py` usage where we can.

```python
if logger is None or isinstance(logger, six.string_types):
```

nicoddemus commented on _pytest/logging.py:
```diff
@@ -0,0 +1,356 @@
+
+    Defaults to the root logger.
+    """
+    if logger is None or isinstance(logger, py.builtin._basestring):
```
This implementation feels a little weird to me, if someone passes something other than a `str` or `None`, it will return `None` as well. Seems simpler to just forward the call to `getLogger` and let it deal with it:

```python
def get_logger_obj(logger=None):
    return logging.getLogger(logger)
```

In fact is debatable if the function is needed at all.

thisch replied:
I also don't like that this function. All test still run if I remove this function and use 
``logger = logger or getlogger(logger)`` instead.

nicoddemus commented on _pytest/logging.py:
```diff
@@ -0,0 +1,356 @@
+    log_level = get_option_ini(config, setting_name)
+    if not log_level:
+        return
+    if isinstance(log_level, py.builtin.text):
```
You can use `six` here as well

nicoddemus commented on _pytest/logging.py:
```diff
@@ -0,0 +1,356 @@
+
+
+def pytest_configure(config):
+    config.pluginmanager.register(LoggingPlugin(config), 'loggingp')
```
Typo in the plugin name

thisch replied:
No, I suffixed the plugin name with 'p' on purpose, since the _pytest.logging module is automatically registered as a pytest plugin(?) somehow. The name 'logging' is therefore already used. Do you have a better idea than coming up with a plugin name, which does not conflict with the name of the python module?

nicoddemus replied:
Ah OK, thanks for the explanation.

That's how pytest works, `_pytest.logging` is already a plugin, and inside that plugin you are registering another plugin. We have the same in `_pytest.capture`, where it installs a `CaptureManager` plugin as `"capman"`.

I suggest using just `logging-plugin` for consistency with `LoggingPlugin`.

nicoddemus commented on _pytest/logging.py:
```diff
@@ -0,0 +1,356 @@
+            if log_cli_level is None:
+                # No log_level was provided, default to WARNING
+                log_cli_level = logging.WARNING
+        self.log_cli_level = log_cli_level
```
I suggest changing `get_actual_log_level` to receive a list of options, and it returns the first non-`None`. This way this entire block becomes:

```python
self.log_cli_level = get_actual_log_level(config, 'log_cli_level', 'log_level') or logging.WARNING
```

The same is valid for `get_option_ini`.

thisch replied:
Nice, this for this suggestion, this works and simplifies the code a bit.

nicoddemus commented on testing/logging/test_fixture.py:
```diff
@@ -0,0 +1,73 @@
+logger = logging.getLogger(__name__)
+sublogger = logging.getLogger(__name__+'.baz')
+
+u = (lambda x: x.decode('utf-8')) if sys.version_info < (3,) else (lambda x: x)
```
I don't think you need this, you can just use plain unicode literals (use `u'bū'` instead of `u('bū')`). This code was needed for Python <3.3 which reintroduced the `u` prefix for unicode literals.

nicoddemus replied:
Btw just to be clear: I know you just ported the code and didn't wrote that bit. My suggestions are small improvements before merging the code. 😁

nicoddemus commented on testing/test_capture.py:
```diff
@@ -344,10 +344,7 @@ def teardown_module(function):
 
     def test_logging_initialized_in_test(self, testdir):
         p = testdir.makepyfile("""
-            import sys
```
Hmm I fear if we will reintroduce the bug described in 2e80512bb8a04f141a34fc6586b3518774954193. It was so long ago (issue number 8) that it probably is fine now.

If we decide this is no longer relevant, we should just remove this test then.

thisch replied:
Who can tell if this is still relevant? Probably the author of the fix (@hpk42)

nicoddemus replied:
If nobody manifests themselves, I propose just removing the test. I don't see how to keep the test around if we introduce the logging module into the core, after all.

thisch commented on _pytest/logging.py:
```diff
@@ -0,0 +1,335 @@
+        if logger is None:
+            logger = self.handler
+        else:
+            logger = logging.getLogger(logger)
```
We should remove the ```else: ..`` code and enforce that the logger kwarg is either None or an instance of a logger object.

nicoddemus replied:
But that would break the current interface, the `logger` parameter currently is a str/name. I would like to avoid breaking the current API so users that were using `pytest-catchlog` can use the builtin logging without changes if possible.

thisch replied:
Okay, then let's leave it as is.

nicoddemus commented on changelog/2794.feature:
```diff
@@ -0,0 +1 @@
+Merge the catchlog plugin.
```
We should make a more expressive announcement, this is a major feature of the new version and most users will only know about this feature from this log entry. 😁 

My suggestion:

```rst
Pytest now captures and displays output from the standard `logging` module. The user can control the logging level to be captured by specifying options in ``pytest.ini``, the command line and also during individual tests using markers. Also, a ``caplog`` fixture is available that enables users to test the captured log during specific tests (similar to ``capsys`` for example). For more information, please see the `docs <https://docs.pytest.org/en/latest/logging.html>`_.

This feature was introduced by merging the popular `pytest-catchlog <https://pypi.org/project/pytest-catchlog/>`_ plugin, thanks to `Thomas Hisch <https://github.com/thisch>`_. Be advised that during the merging the backward compatibility with the defunct ``pytest-capturelog`` has been dropped.
```

nicoddemus commented on doc/en/usage.rst:
```diff
@@ -189,6 +189,187 @@ in your code and pytest automatically disables its output capture for that test:
 
+.. _logging:
+
+Logging
```
I believe this live in its own file, `logging.rst`.

nicoddemus commented on doc/en/usage.rst:
```diff
@@ -189,6 +189,187 @@ in your code and pytest automatically disables its output capture for that test:
+Live Logs
+^^^^^^^^^
+
+By default, catchlog will output any logging records with a level higher or
```
Please make a search for the name ``catchlog`` and replace it with ``pytest``.

nicoddemus commented on doc/en/usage.rst:
```diff
@@ -189,6 +189,187 @@ in your code and pytest automatically disables its output capture for that test:
   and continue with the regular test run.
 
+.. _logging:
+
```
If a user has ``pytest-catchlog`` installed what will happen? Probably they will step in each other's toes.

Perhaps we should put a note here:

```rst
... note::
    This feature is a drop-in replacement for the `pytest-catchlog <https://pypi.org/project/pytest-catchlog/>`_ plugin
    and they will conflict with each other. The backward compatibility API with ``pytest-capturelog`` has been dropped 
    when this feature was introduced, so if for that reason you still need ``pytest-catchlog`` you can disable the internal
    feature by adding to your ``pytest.ini``:

    .. code-block:: ini

        [pytest]
        addopts=-p no:logging
```

thisch commented:
I added a logging.Formatter class, which colors the loglevel and supports multiline log strings, in the most recent commit. The reason why I added it is that i love colored log output ;). I'm fine if you don't want this Formatter to be part of this PR.  WDYT in general, should we color the (CLI) log loglines?

nicoddemus commented on _pytest/logging.py:
```diff
@@ -0,0 +1,423 @@
+                     '%(colored_levelname)8s ' +
+                     self.separator +
+                     '%(message)s')
+        try:
```
Just a minor comment: I usually avoid try/except/ImportError like this because an `ImportError` might mean that the module does not exist, but might be caused by some other problem (like a missing dependency). I explain this in more details in pytest-dev/pytest-mock#68, that was caused by a problem in the installed `mock` package, where a dependency was missing and it was raising an `ImportError`.

Like I said just a minor comment, no need to change this given that  `logging` is builtin. 😁

thisch replied:
I had an explict check for sys.version in a previous version of the logcolor commit, but decided to switch to the try/except/importerror code, since I thought that this is more pythonic ;). In general I agree with you, but since a module from the stdlib is imported, I doubt that the problems you mentioned can occur.

nicoddemus commented:
> I added a logging.Formatter class, which colors the loglevel and supports multiline log strings, in the most recent commit. The reason why I added it is that i love colored log output ;).

I also like colored output, so I'm 👍 on this. Can you perhaps post a screenshot on how it looks like so others can easily comment on it?

thisch commented:
Sure, (An additional newline is missing after each testname. I don't yet know how I can achieve this, though.)

![image](https://user-images.githubusercontent.com/206581/30936361-f6b4e5d6-a3d3-11e7-83a3-4e4f6d118055.png)

nicoddemus commented:
Oh sorry, I meant from a pytest run which captures and shows some logs. 😁

thisch commented:
![image](https://user-images.githubusercontent.com/206581/30937518-54b8c9ce-a3d7-11e7-9ef6-8c5de36d8443.png)
Is this better? :)

nicoddemus commented:
It looks great!

One question, why is the colored log under "Captured stderr call"? Shouldn't it be under "Captured log call"?

thisch commented:
Good question, I'll have a look tomorrow.

nicoddemus commented on _pytest/logging.py:
```diff
@@ -0,0 +1,423 @@
+
+        self.separator = '| '
+        self.datefmt = '%H:%M:%S.%f'
+        self._colored_levels = {
```
set comprehensions are not supported in python 2.6 😞

RonnyPfannschmidt replied:
#2812 might elevate that

thisch replied:
BTW If you don't mind, I'll create a separate PR for the Colored Log Formatter and remove the Formatter from this commit (there are some details which need to be settled first).

RonnyPfannschmidt replied:
that should work out fine, thanks :)

nicoddemus commented:
@thisch anything else besides removing `test_logging_initialized_in_test` and updating the changelog that you still want to do here?

From my POV after those two issues are dealt with we can merge this and make a 3.3 release. 👍

nicoddemus commented:
Thanks @thisch! You want me to update the CHANGELOG? It is easy to do so from GH's UI if you would like me to.

thisch commented:
I squashed your suggested changelog entry into the 'Add changelog ...' commit

nicoddemus commented:
Oh thanks I didn't notice that! 👍 

LGTM, thanks again for tackling this awesome new feature.

thisch commented:
I did this a few minutes ago.

Note that the performance tests are still not part of this PR. I don't plan on merging them into pytest, though.

The-Compiler commented:
@thisch If you have the time, can you please take a look at the remaining pytest-catchlog issues/PRs? See  https://github.com/pytest-dev/pytest/issues/3003. I also opened https://github.com/eisensheng/pytest-catchlog/pull/73 to make it clear what's going on.

RonnyPfannschmidt commented on _pytest/logging.py:
```diff
@@ -0,0 +1,337 @@
+    """Add options to control log capturing."""
+    group = parser.getgroup('logging')
+
+    def add_option_ini(option, dest, default=None, type=None, **kwargs):
```
while reiterating this as it got back into my notifications i beleive this one should have never met a release,
literally all the options we add are technically superseeded by the much better option setting argument

nicoddemus replied:
I see what you mean, all command-line options could be replaced by using the `-o` option instead. 

While I agree with your reasoning, we have to remember that one of the objectives was to merge this plugin as is to allow a drop-in-replacement for users, so removing these options would break their usage.
